auto_compressor/src/state_saving.rs

```diff
@@ -0,0 +1,249 @@
+        // Mark the previous level_number seen as the current one
+        prev_seen = level_num;
+    }
+
+    // If we didn't retrieve anything from the database then there is no saved state
+    // in the database!
+    if level_info.is_empty() {
+        return Ok(None);
+    }
+
+    // Return the compressor state we retrieved
+    // last_compressed cannot be None at this point, so safe to unwrap
+    Ok(Some((last_compressed.unwrap(), level_info)))
+}
+
+/// Save the level info so it can be loaded by the next run of the compressor
+///
+/// # Arguments
+///
+/// * `client`            - A postgres client used to send the requests to the database
+/// * `room_id`           - The room who's saved compressor state we want to save
+/// * `level_info`        - The state that can be used to restore the compressor later
+/// * `last_compressed`   - The last state_group that was compressed. This is needed
+///                         so that the compressor knows where to start from next
+pub fn write_room_compressor_state(
+    client: &mut Client,
+    room_id: &str,
+    level_info: &[Level],
+    last_compressed: i64,
+) -> Result<()> {
+    // Wrap all the changes to the state for this room in a transaction
+    // This prevents accidentally having malformed compressor start info
+    let mut write_transaction = client.transaction()?;
+
+    // Go through every level that the compressor is using
+    for (level_num, level) in level_info.iter().enumerate() {
+        // the 1st level is level 1 not level 0, but enumerate starts at 0
+        // so need to add 1 to get correct number
+        let level_num = level_num + 1;
+
+        // bring the level info out of the Level struct
+        let (max_size, current_len, current_head) = (
+            level.get_max_length(),
+            level.get_current_length(),
+            level.get_head(),
+        );
+
+        // Update the database with this compressor state information
+        //
+        // Some of these are `usize` as they may be used to index vectors, but stored as Postgres
+        // type `INT` which is the same as`i32`.
+        //
+        // Since these values should always be small, this conversion should be safe.
+        let (level_num, max_size, current_len) =
+            (level_num as i32, max_size as i32, current_len as i32);
+        let params: Vec<&(dyn ToSql + Sync)> =
+            vec![&room_id, &level_num, &max_size, &current_len, &current_head];
+
+        write_transaction.execute(
+            r#"
+                INSERT INTO state_compressor_state 
+                    (room_id, level_num, max_size, current_length, current_head) 
+                    VALUES ($1, $2, $3, $4, $5)
+                ON CONFLICT (room_id, level_num) 
+                    DO UPDATE SET (max_size, current_length, current_head) 
+                        = (excluded.max_size, excluded.current_length, excluded.current_head);
```

Comment:
I think this uses the norm UPDATE syntax,  so:

```suggestion
                    DO UPDATE SET
                        max_size = EXCLUDED.max_size,
                        current_length = EXCLUDED.current_length,
                        current_head = EXCLUDED.current_head
```

?

Reply:
that work!